pkg/runtime/grpc/k8s-portfwd-dialer.go

```diff
@@ -0,0 +1,154 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcruntime
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/kubectl-gadget/utils"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/factory"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+)
+
+type k8sPortFwdDialer struct {
+	io.Writer
+	io.Reader
+	conn    httpstream.Connection
+	stream  httpstream.Stream
+	podName string
+}
+
+// NewK8SPortFwdConn connects to a Pod using PortForwarding via the Kubernetes API Server
+func NewK8SPortFwdConn(ctx context.Context, pod target, timeout time.Duration) (net.Conn, error) {
+	conn := &k8sPortFwdDialer{}
+
+	config, err := utils.KubernetesConfigFlags.ToRESTConfig()
+	if err != nil {
+		return nil, fmt.Errorf("creating RESTConfig: %w", err)
+	}
+
+	// set GroupVersion and NegotiatedSerializer for RESTClient
+	factory.SetKubernetesDefaults(config)
+
+	conn.podName = pod.addressOrPod
+
+	config.Timeout = timeout
+
+	transport, upgrader, err := spdy.RoundTripperFor(config)
+	if err != nil {
+		return nil, fmt.Errorf("creating roundtripper: %w", err)
+	}
+
+	targetURL, err := url.Parse(config.Host)
+	if err != nil {
+		return nil, fmt.Errorf("parsing restConfig.Host: %w", err)
+	}
+
+	targetURL.Path = path.Join(
+		"api", "v1",
+		"namespaces", "gadget",
+		"pods", conn.podName,
+		"portforward",
+	)
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, targetURL)
+
+	newConn, _, err := dialer.Dial(portforward.PortForwardProtocolV1Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// create error stream
+	headers := http.Header{}
+	headers.Set(v1.StreamType, v1.StreamTypeError)
+	headers.Set(v1.PortHeader, fmt.Sprintf("%d", api.GadgetServicePort))
+	headers.Set(v1.PortForwardRequestIDHeader, strconv.Itoa(1))
+	errorStream, err := newConn.CreateStream(headers)
+	if err != nil {
+		newConn.Close()
+		return nil, fmt.Errorf("creating error stream for port forward: %w", err)
+	}
+	// we're not writing to this stream, but it is required for other streams to be able to connect
+	errorStream.Close()
+
+	go func() {
+		message, err := io.ReadAll(errorStream)
+		switch {
+		case err != nil:
+			log.Errorf("k8sPortFwd connection: reading from error stream: %v", err)
+		case len(message) > 0:
+			log.Errorf("k8sPortFwd connection: forwarding port: %v", string(message))
+		}
```

Comment:
> Shouldn't this code run in a loop?

In theory we could, but we don't really need that information in realtime (the endpoint could handle several port forwarding connections over time, but we just use one), so there's no need to handle that protocol in more detail here. We still need to subscribe and handle that stream, though, otherwise forwarding does not work.

Reply:
The `Close()` just closes the sending direction of that stream, while a `Reset()` would close both directions afaiu. Naming is a bit confusing there.